src/Compilers/CSharp/Portable/Binder/Binder_Initializers.cs

```diff
@@ -150,44 +151,44 @@ internal struct ProcessedFieldInitializers
                     }
 
                     var syntaxRef = initializer.Syntax;
-                    Debug.Assert(syntaxRef.SyntaxTree.Options.Kind != SourceCodeKind.Regular);
+                    var syntaxTree = syntaxRef.SyntaxTree;
+                    Debug.Assert(syntaxTree.Options.Kind != SourceCodeKind.Regular);
 
-                    var initializerNode = (CSharpSyntaxNode)syntaxRef.GetSyntax();
+                    var syntax = (CSharpSyntaxNode)syntaxRef.GetSyntax();
+                    var syntaxRoot = (CSharpSyntaxNode)syntaxTree.GetRoot();
 
                     if (binderFactory == null)
                     {
-                        binderFactory = compilation.GetBinderFactory(syntaxRef.SyntaxTree);
+                        binderFactory = compilation.GetBinderFactory(syntaxTree);
                     }
 
-                    Binder scriptClassBinder = binderFactory.GetBinder(initializerNode);
-                    Debug.Assert(((ImplicitNamedTypeSymbol)scriptClassBinder.ContainingMemberOrLambda).IsScriptClass);
+                    var scriptClassBinder = binderFactory.GetBinder(syntax);
+                    Debug.Assert(((NamedTypeSymbol)scriptClassBinder.ContainingMemberOrLambda).IsScriptClass);
 
                     if (firstDebugImports == null)
                     {
                         firstDebugImports = scriptClassBinder.ImportChain;
                     }
 
-                    Binder parentBinder = new ExecutableCodeBinder((CSharpSyntaxNode)syntaxRef.SyntaxTree.GetRoot(), scriptInitializer, scriptClassBinder);
+                    var parentBinder = new ExecutableCodeBinder(syntaxRoot, scriptInitializer, scriptClassBinder).GetBinder(syntaxRoot);
```

Comment:
GetBinder will return a ScriptInitializerBinder since it is called with the CompilationUnitSyntax. See LocalBinderFactory.VisitCompilationUnit.